expression/expression.go

```diff
@@ -110,53 +110,54 @@ func (t ComparisonExpression) String() string {
 	return fmt.Sprintf("(%s %s %s)", l, t.Comparison, r)
 }
 
-/* Logical Node
-
-type LogicalNode struct {
-	Left    Node
-	Right   Node
+// Logical expressions
+// Contains expressions such as a Or b, a And b
+type LogicalExpression struct {
+	Left    Expression
+	Right   Expression
 	Logical Logical
-	Objects []Object
 }
 
-func (t LogicalNode) Evaluate(ctx Context) (interface{}, error) {
+func (t LogicalExpression) Evaluate(ctx Context) (interface{}, error) {
 	if t.Left == nil {
 		return nil, errors.New("Left node is nil")
 	}
+
 	if t.Right == nil {
 		return nil, errors.New("Right node is nil")
 	}
 
-	left, err := t.Left.Evaluate(tracer, ctx)
+	left, err := t.Left.Evaluate(ctx)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Left logical evaluate failed")
+		return nil, errors.Wrap(err, "Left logical evaluate failed")
 	}
 
-	right, err := t.Right.Evaluate(tracer, ctx)
+	right, err := t.Right.Evaluate(ctx)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Right logical evaluate failed")
+		return nil, errors.Wrap(err, "Right logical evaluate failed")
+	}
+
+	if left == nil || right == nil {
+		return nil, nil
+
 	}
 
-	var result bool
+	// Making this a switch statement so that we can add more operations later
 	switch t.Logical {
-	case And:
-		result, err = OperationAnd(left, right)
 	case Or:
-		result, err = OperationOr(left, right)
+		return OperationOr(left, right)
 	default:
-		result, err = false, LogicalError
+		return nil, LogicalError
 	}
 
-	return evaluateResult(tracer, result, err, t.Objects)
+	return nil, nil
 }
 
-func (t LogicalNode) String() string {
-	return nodeToString(t.Left, t.Right, t.Logical)
+func (t LogicalExpression) String() string {
+	return fmt.Sprintf("%s %s %s", t.Left, t.Logical, t.Right)
```

Comment:
Look at the comparison String() method, this maybe crash if there is a nil value